Customizable Bars (Health, Shield and much more)

Fryone opened the pull request:
Added customizable bars in form of BarTypes.

Example
```ini
[BarTypes] ; section in rulesmd.ini
0=HEALTHBAR1

[HEALTHBAR1]
InfoType= ;some infotype
PipBrd.Background.File= ; filename - without the .shp extension, image for use as background, first frame
PipBrd.Background.ShowWhenNotSelected=false ; boolean, should background be drawn, when ponting on techno
PipBrd.Background.Translucency= ; translucency level (0/25/50/75), background image translucency level
PipBrd.Foreground.File= ; filename - without the .shp extension, image for use as foreground (drawn after health sections), first frame
PipBrd.Foreground.ShowWhenNotSelected=false ; boolean, should foreground be drawn, when ponting on techno
PipBrd.Foreground.Translucency= ; translucency level (0/25/50/75), foreground image translucency level
PipBrd.Offset=0,0 ; X,Y, offset of background and foreground boards images
Bar.Offset=0,0 ; X,Y, offset of whole bar
Pips.File= ; filename - without the .shp extension, image to take frames from
Pips.Frames=16,17,18 ; list of 3 integers, frames of pips.shp to use (green, yellow, red states)
Pips.EmptyFrame=-1 ; integer, frame in pips.shp to use for empty section
Pips.Amount=17 ; integer, amount of sections, when bar is full
Pips.PositionDelta=2,0 ; X,Y, increment in pixels to next section
Pips.DrawBackwards=false ; boolean, changes order of drawing from last to first

[AudioVisual]
Buildings.DefaultBarTypes= ;list of bar types
Infantry.DefaultBarTypes= ;list of bar types
Vehicles.DefaultBarTypes= ;list of bar types
Aircraft.DefaultBarTypes= ;list of bar types

[SOMETECHNO]
BarTypes= ;list of bar types
```

No docs for now, needs testing.
![image](https://github.com/user-attachments/assets/016033d5-4f13-4365-8ce5-55b51e321d7a)

Fryone committed:
custom health bar types

Initial commit

tyuah8 commented:
> Added customizable vertical health bars in form of HealthBarTypes. By default, technos use their own vanilla healthbars. Adding `CustomHealthBarType` changing health bar to this type. Can be applied to buildings.
> 
> Example
> 
> ```
> [HealthBarTypes]                                                       ; section in rulesmd.ini
> 0=HEALTHBAR1
> 
> [HEALTHBAR1]
> Frame.Background=PIPBRD                                      ; shp file for use as background, first frame
> Frame.Background.Translucency=0                           ; background translucency level
> Frame.Background.ShowWhenNotSelected=false    ; should background be drawn, when ponting on techno
> Frame.Foreground=                                                   ; shp file for use as foreground (drawn after health sections), first frame
> Frame.Foreground.ShowWhenNotSelected=false     ; should foreground be drawn, when ponting on techno
> HealthBar.OffsetX=0                                                  ; horizontal offset, similar to PixelSelectionBracketDelta
> Sections.Pips=16,17,18                                              ; frames of pips.shp to use (green, yellow, red states)
> Sections.Amount=17                                                 ; amount of sections, when bar is full
> Sections.Size=2                                                          ; size in px of individual section
> 
> [SOMETECHNO]
> CustomHealthBarType=HEALTHBAR1
> ```
> 
> No docs for now, needs testing. ![image](https://private-user-images.githubusercontent.com/30149531/370817781-016033d5-4f13-4365-8ce5-55b51e321d7a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczMjIwMDYsIm5iZiI6MTcyNzMyMTcwNiwicGF0aCI6Ii8zMDE0OTUzMS8zNzA4MTc3ODEtMDE2MDMzZDUtNGYxMy00MzY1LThjZTUtNTViNTFlMzIxZDdhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDAzMzUwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjNDE1N2JkMzQ2NTQ3Y2E4NzJiMDcwYjAxZDQxMGM5OTNhMzUwNDdmMjc2MjczMTBkY2QyY2MwNDE4ZWE4N2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FnAm-3GtGKGuXp90t5Ssqc-Mznp9W__E2bEMrF8J93E)

Those voxels are quite nice and seem to resemble the look and feel of the original game voxels.  Did you make those voxels yourself?

Fryone commented:
> Those voxels are quite nice and seem to resemble the look and feel of the original game voxels. Did you make those voxels yourself?

yep

NetsuNegi commented:
nice feature for epic vehicle/airship/battleship :)

Metadorius commented:
![image](https://github.com/user-attachments/assets/c69d858e-00a5-41f8-9572-6db0b770a6f1)
Custom Health Bar

Metadorius commented:
Couple of suggestions on the general API stuff:
`[HealthBarTypes]` => `[BarTypes]`
In general I propose to treat this as a "custom anything bar" which can be used for HP bars.

For technos the tag should be `CustomHealthBarType` => `HealthBar`, less redundancy and more in-line with other things.

I think shields should use that too.

Metadorius commented:
P.S. Can you do any shape/direction bars this way? I think it's a matter of making the section offset customizable in two dimensions, this way you could specify the 2,2 offset and have the bar go diagonal for example.

Fryone commented:
> P.S. Can you do any shape/direction bars this way? I think it's a matter of making the section offset customizable in two dimensions, this way you could specify the 2,2 offset and have the bar go diagonal for example.

The horizontal (vanilla like) bars were first in mind. I understand vertical ones, but diagonal...

Metadorius commented:
> The horizontal (vanilla like) bars were first in mind. I understand vertical ones, but diagonal...

It's only an example. The point is that if one can specify offset between sections in 2 dimensions then they can come up with whatever weird graphics they want, including diagonal bars, sectioned graphics etc.

Fryone commented:
I think, I probably should make some "BarMaker" then, like procedure to draw by position and type.

chaserli commented:
#1287 #1309 #1395
https://github.com/Phobos-developers/Phobos/pull/1287#issuecomment-2198477945
If you are making a bar, generalize it to cover all fractionable quantities. HP/SP are just a small part of them. Some should display regardless of mouse hovering/selected.

Fryone commented:
> #1287 #1309 #1395 [#1287 (comment)](https://github.com/Phobos-developers/Phobos/pull/1287#issuecomment-2198477945) If you are making a bar, generalize it to cover all fractionable quantities. HP/SP are just a small part of them. Some should display regardless of mouse hovering/selected.

yeah, so the bars can be used to draw without hovering/selecting. i'm just using bar-drawer in the vanilla health part code. So it behaves like vanilla health bar.

chaserli commented:
> yeah, so the bars can be used to draw without hovering/selecting. i'm just using bar-drawer in the vanilla health part code. So it behaves like vanilla health bar.

I mean you can go more aggressive than that, just replace all 4 DrawHealthBar vtbl calls and handle it as well as other things in drawextras.

Fryone committed:
changed to BarTypes

- added emptypip
- added draw direction
- added border offset
- changed overall logic

ZivDero commented on src/New/Type/BarTypeClass.h:
```diff
@@ -0,0 +1,48 @@
+class BarTypeClass final : public Enumerable<BarTypeClass>
+{
+public:
+	Nullable<SHPStruct*> BoardBG_File;
+	Valueable<bool> BoardBG_ShowWhenNotSelected;
+	Nullable<TranslucencyLevel> BoardBG_Translucency;
+	Nullable<SHPStruct*> BoardFG_File;
+	Valueable<bool> BoardFG_ShowWhenNotSelected;
+	Nullable<TranslucencyLevel> BoardFG_Translucency;
+	Valueable<Vector2D<int>> Board_Offset;
```
Board -> Frame?

also BG/FG -> Background/Foreground, please

Metadorius replied:
Not sure why translucency flags are nullable here when they can just be valueables with default 0?

Metadorius commented on src/Ext/Techno/Body.Visuals.cpp:
```diff
@@ -469,3 +469,62 @@ void TechnoExt::GetValuesForDisplay(TechnoClass* pThis, DisplayInfoType infoType
+		}
+		
+		position = pLocation;
+		position.X += drawBackwards ? (int)round(sectionAmount * sectionOffset.X / 2) : -(int)round(sectionAmount * sectionOffset.X / 2);
+		position.Y += drawBackwards ? (int)round(sectionAmount * sectionOffset.Y / 2) : -(int)round(sectionAmount * sectionOffset.Y / 2);
```
1. Pretty sure you can do math with vectors like normal, you don't need to duplicate it for each coord.
2. Instead of a ternary operator and duplication everywhere where you need to decide on the direction of draw you can do `int sign = drawBackwards * -2 + 1;` and multiply by that every time you used a ternary here.

Metadorius replied:
This all applies to multiple lines in the code btw, forgot to comment on all of them, please see where it is applicable because there's a lot of duplication going on.

Metadorius commented on src/Ext/Techno/Body.Visuals.cpp:
```diff
@@ -469,3 +469,62 @@ void TechnoExt::GetValuesForDisplay(TechnoClass* pThis, DisplayInfoType infoType
+		for(int i = 0; i < sectionAmount; ++i)
+		{
+			position += drawBackwards ? -sectionOffset : sectionOffset;
+			DSurface::Temp->DrawSHP(FileSystem::PALETTE_PAL, FileSystem::PIPS_SHP(),
```
Could make `pips.shp` customizable too?

ZivDero commented:
Also, you're hooking into the function. But the vanilla function is just one big if-else, and you hook both of the branches. You should replace the entire function.

Additionally, your hooks disable other Phobos hooks (shield bars, HeathBar.Height, digital displays). As such, you should move those patches into your replaced function, too.

Metadorius commented:
@Fryone is there anything blocking you from fulfilling the requested changes?

Fryone commented:
> @Fryone is there anything blocking you from fulfilling the requested changes?

Nothing,  I just need to get it. I just need to do it.

Fryone committed:
Update src/Ext/Techno/Body.Visuals.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Update src/Ext/Techno/Body.Visuals.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Update src/Ext/Techno/Hooks.Pips.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Update src/Ext/Techno/Hooks.Pips.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Update src/Ext/Techno/Body.Visuals.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Update src/New/Type/BarTypeClass.h

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Update src/Ext/Techno/Body.Visuals.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Update src/Ext/Techno/Hooks.Pips.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Update src/Ext/Techno/Hooks.Pips.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Update src/Ext/Techno/Hooks.Pips.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Fryone committed:
Rework 1

- hook refactor
- Healthbar.Hide moved to healthbar hook
- applied some suggested changes
- added ability to choose file for bar pips
- added ShieldBar.BarType

TO DO:
- for now, ShieldBar.BarType dont work without HealthBar.BarType

Fryone committed:
Merge remote-tracking branch 'upstream/develop' into fryone/healthbartypes

Fryone committed:
rework update

- changed blitterflags
- updated logic with latest hook movements
- shield and health bars can be used separately

Metadorius commented:
@Fryone any response on the comments?

Fryone commented:
I'm thinking if I should just finish whats done without any new logic changes (which is yrpp constants and docs)

Metadorius commented:
> I'm thinking if I should just finish whats done without any new logic changes (which is yrpp constants and docs)

On the `InfoType`: while I am not sure which approach is better, I do agree that it would be better to have this part common between digital displays and bars. Essentially, bars and digital displays do the same thing but in a bit different way, so they should be similar in use.

From what I understood, @CrimRecya (and previously @chaserli too) suggest a unified way to draw all such indicators and a "framework" of sorts.

CrimRecya commented:
If possible, I also believe that #1422 is the best solution. But it requires a lot of effort, so I would like to ask for your opinion. Can you help complete this framework? if you don't have that much time either, perhaps we can also consider merging this pr first and modifying it in the future and using the migration document.

Fryone commented:
the idea:

[InfoDisplayTypes]
DisplayFormatType=        ; BarType / DigitalDisplayType
InfoType=                 ; Displayed value enumeration (health|shield|ammo|mindcontrol|spawns|passengers|tiberium|experience|occupants|gattlingstage)
_//all positioning and visibility related stuff//_

CrimRecya commented:
I prefer to use separated `BarType` and `DigitalDisplayType`.
`BarType` may require some other drawing method, while `DigitalDisplayType` does not;
`DigitalDisplayType` may need to display like a timer, while `BarType` does not.
It's just that their data sources are all from `InfoType`.

Fryone committed:
Merge remote-tracking branch 'upstream/develop' into fryone/healthbartypes

NetsuNegi commented:
Considering that the vanillia filename is called "pips.shp", "pipbrd.shp", maybe it's better to use `PipBrd` and `Pips` instead of `BoardBG` and `Section`

Metadorius commented:
> from design perspective #1635 is clearer and more universal

Can you list what specifically needs improvement with examples? I think @Fryone could incorporate @NetsuNegi's improvements over this PR, or @NetsuNegi could submit some changes as he has edit rights as a maintainer.

Fryone commented:
I already changed tag names, just not uploaded it here. Plus, there is something that using any shield bar type would crash game.

Fryone committed:
Merge remote-tracking branch 'upstream/develop' into fryone/healthbartypes

Fryone committed:
latest state

- changed logic and namings
- shields does not work
- some problems with value calculations

Fryone commented:
Planned: 
1) Proper calculations; 
2) Adding "DrawEmptyBar" for some infotypes;
3) Adding support for new infotypes;
4) Fixing bars for shields.

A1026K78B commented:
Here are some ideas for the custom health bar mod:

1.Multi-layer health bar.Like this.
<img width="247" height="82" alt="image" src="https://github.com/user-attachments/assets/7dfe1aca-60b9-4393-bdc8-8a20c210bc3c" />
<img width="240" height="69" alt="image" src="https://github.com/user-attachments/assets/e7ae8fac-dd1e-4d64-835d-4a2a001e3d4d" />

2.Add light-colored pips that temporary appear for a few frames when taking damage, indicating the amount of damage taken per hit.
<img width="423" height="154" alt="image" src="https://github.com/user-attachments/assets/3b79acd7-0bde-4f72-9cd7-68243b551fb1" />

Fryone committed:
Merge remote-tracking branch 'upstream/develop' into fryone/healthbartypes

Fryone committed:
update renewed 1

shield check fix
yrpp update

Fryone committed:
update renewed 1.1

Fryone committed:
update renewed 2

- removed hook and used similar to DigitalDisplay procedure
- values check used from DigitalDisplay
 * vanilla and new health bars behave a bit different
- no info index atm

Fryone committed:
update renewed 3

 - added InfoIndex
 - added Bar.ConditionYellow and Bar.ConditionRed

Fryone committed:
update renewed 4

- added Pips.ChangePerSection
   * changes pip frame only for the last section
